File: Algorithms/Challenges/Strings/LexicographicRotation/Java/LexicographicRotation.java
import java.util.*;
import java.io.*;

/*
Description
Given a string return the lexicographic maximum/minimum rotation

Example:

Input: input_string
rotate

Output: output_rotation
terota
*/

class LexicographicRotation {

  static String[] allRotations(String str) {
    int n = str.length();
    String[] arr = new String[n];
    String comboStr = str + str;
    for (int i = 0; i < n; i++) arr[i] = comboStr.substring(i, i+n);
    return arr;
  }

  static String getMaxRotation(String str) {
    String[] rotations = allRotations(str);
    Arrays.sort(rotations);
    return rotations[rotations.length - 1];
  }

  static String getMinRotation(String str) {
    String[] rotations = allRotations(str);
    Arrays.sort(rotations);
    return rotations[0];
  }

  static String solve(String str) {
    String[] params = str.split(" ");
    String inputStr = params[0];
    int i = Integer.parseInt(params[1]);
    if (i == 0) return getMinRotation(inputStr);
    if (i == 1) return getMaxRotation(inputStr);
    return "Invlid Input";
  }

  public static void main(String[] args) throws IOException, FileNotFoundException {
    BufferedReader br = new BufferedReader(new FileReader("io/tests.txt"));
    BufferedWriter bw = new BufferedWriter(new FileWriter("io/solution.txt"));
    boolean debug = true;

    LexicographicRotation classInstance = new LexicographicRotation(); // If non-static methods

    String test;
    while ( (test = br.readLine()) != null ) {
      String solution = String.valueOf(solve(test));
      if (debug) solution +=  ": '" + test + "'";
      bw.write(solution);
      bw.newLine();
    }

    br.close();
    bw.close();
  }
}

File: Algorithms/Challenges/Strings/LexicographicRotation/Python/LexicographicRotation.py
"""
Description
Given a string return the lexicographic maximum/minimum rotation

Example:

Input: input_string int (0 = min, 1 = max)
rotate 1

Output: output_rotation
terota
"""

def allRotations(string):
    n = len(string)
    arr = []
    combo_str = string + string
    for i in range(n):
        arr.append(combo_str[i:i+n])
    return arr

def getMaxRotation(string):
    rotations = allRotations(string)
    rotations.sort()
    return rotations[-1]

def getMinRotation(string):
    rotations = allRotations(string)
    rotations.sort()
    return rotations[0]

def solve(input):
    params = input.split(" ")
    string = params[0]
    i = int(params[1])
    if i == 0: return str(getMinRotation(string))
    if i == 1: return str(getMaxRotation(string))
    return "Invlid Input"

input_file = open("io/tests.txt", "r");
output_file = open("io/solution.txt", "w");
debug = True

test = input_file.readline()
while test != '':
    test = test.rstrip("\r\n")
    solution = solve(test)
    if debug: solution +=  ": '%s'" % test
    output_file.write(solution + "\n")
    test = input_file.readline()

input_file.close();
output_file.close();
